protocols/kad/src/behaviour.rs

```diff
@@ -732,72 +770,114 @@ where
                                 target = kbucket::Key::new(PeerId::random());
                             }
                             target
-                        }).collect::<Vec<_>>();
+                        }).collect::<Vec<_>>().into_iter();
```

Comment:
Why not make this iterator lazy delaying the _heavy_ computations to the time they are needed.

Reply:
What exactly do you have in mind? Note the following:

  * The absolute maximum size of this vector is 254 elements, but this is _extremely_ unlikely (probabilistically). The realistic sizes are 1-10 elements.
  * There are no heavy computations (at least I don't consider hashing, xor and sampling bytes from a PRNG to be heavy computations).
  * We need to know the length of this vector / iterator immediately below, so if we wanted to leave it as an iterator, it must be an `ExactSizeIterator`.
  * `remaining` is put into the `QueryInfo` for which I'd like to avoid adding type parameters.

Nevertheless, please let me know if you have a concrete approach in mind that is not invasive, maybe I'm missing something!